File: topos/mcp/tools/compare.py
"""
Structural comparison tools: AST edit distance between two programs.
"""

from __future__ import annotations

from fastmcp.tools.base import ToolResult

from topos.core.morphism import ProgramMorphism
from topos.functors.profunctors.ast.compare import calculate_ast_distance

from ..formatting import to_tool_result
from ..schemas import (
    CompareCodeInput,
    CompareFilesInput,
    ComparisonResult,
)
from ..security import read_safe_utf8_file
from ..server import mcp

_READ_ONLY_ANN = {
    "title": "Topos Structural Comparison",
    "readOnlyHint": True,
    "destructiveHint": False,
    "idempotentHint": True,
    "openWorldHint": False,
}


def _failed_comparison(
    error: str, source_valid: bool = False, target_valid: bool = False
) -> ComparisonResult:
    return ComparisonResult(
        raw_distance=0.0,
        normalized_distance=0.0,
        similarity=0.0,
        operations={},
        source_valid=source_valid,
        target_valid=target_valid,
        error=error,
    )


@mcp.tool(
    name="topos_compare_code",
    tags={"compare"},
    annotations=_READ_ONLY_ANN,
)
def topos_compare_code(params: CompareCodeInput) -> ToolResult:
    """Compute the AST (tree-edit) distance between two source-code strings.

    Read-only and idempotent; parses both snippets in memory, never writes or
    scores. Use for clone detection or to measure refactor impact; the
    ``topos_assess_*`` tools already fold this in as an anti-gaming check, so
    call it directly only for the raw number. Returns a ComparisonResult:
    ``normalized_distance`` in [0, 1], ``similarity`` (= 1 - it),
    ``raw_distance``, an ``operations`` edit-count map, and
    ``source_valid``/``target_valid`` (``error`` set if either fails to parse).
    """
    try:
        src = ProgramMorphism(source=params.source_code, language=params.language)
        tgt = ProgramMorphism(source=params.target_code, language=params.language)
    except Exception as exc:
        model = _failed_comparison(str(exc))
        return to_tool_result(model, render_comparison_md(model))

    if not (src.is_valid and tgt.is_valid):
        model = _failed_comparison(
            "Failed to parse one or both code snippets.",
            source_valid=src.is_valid,
            target_valid=tgt.is_valid,
        )
        return to_tool_result(model, render_comparison_md(model))

    result = calculate_ast_distance(src.ast, tgt.ast)
    model = ComparisonResult(
        raw_distance=result.raw_distance,
        normalized_distance=result.normalized_distance,
        similarity=1.0 - result.normalized_distance,
        operations=dict(result.operations),
        source_valid=True,
        target_valid=True,
    )
    return to_tool_result(model, render_comparison_md(model))


@mcp.tool(
    name="topos_compare_files",
    tags={"compare"},
    annotations=_READ_ONLY_ANN,
)
def topos_compare_files(params: CompareFilesInput) -> ToolResult:
    """Compute the AST (tree-edit) distance between two source files on disk.

    Read-only; parses both files, never writes or scores. Use for clone
    detection or refactor impact; use ``topos_assess_*`` for a quality verdict.
    Returns a ComparisonResult: ``normalized_distance`` in [0, 1],
    ``similarity`` (= 1 - it), ``raw_distance``, an ``operations`` edit-count
    map, and ``source_valid``/``target_valid`` (``error`` set on read/parse
    failure).
    """
    source_text, source_err = read_safe_utf8_file(params.source)
    if source_err:
        model = _failed_comparison(
            f"Source file error: {source_err['error']}",
            source_valid=False,
            target_valid=False,
        )
        return to_tool_result(model, render_comparison_md(model))
    target_text, target_err = read_safe_utf8_file(params.target)
    if target_err:
        model = _failed_comparison(
            f"Target file error: {target_err['error']}",
            source_valid=True,
            target_valid=False,
        )
        return to_tool_result(model, render_comparison_md(model))
    return topos_compare_code(
        CompareCodeInput(
            source_code=source_text or "",
            target_code=target_text or "",
        )
    )


# ---------------------------------------------------------------------------
# Markdown renderer (rendered into ToolResult.content)
# ---------------------------------------------------------------------------


def render_comparison_md(r: ComparisonResult) -> str:
    """Compact markdown for an AST-distance comparison."""
    if r.error:
        return f"**Error:** {r.error}"
    lines = [
        f"**Normalized distance:** {r.normalized_distance:.3f}",
        f"**Similarity:** {r.similarity:.3f}",
        f"**Raw distance:** {r.raw_distance:.1f}",
        f"**Validity:** source={r.source_valid}, target={r.target_valid}",
    ]
    if r.operations:
        ops = ", ".join(f"{k}={v}" for k, v in sorted(r.operations.items()))
        lines.append(f"**Operations:** {ops}")
    return "\n".join(lines)

File: crates/topos-mcp/src/tools/compare.rs
//! Structural comparison tools: AST edit distance between two programs.

use rmcp::handler::server::wrapper::Parameters;
use rmcp::model::CallToolResult;
use rmcp::{tool, tool_router};
use topos_core::core::morphism::ProgramMorphism;
use topos_core::functors::profunctors::ast::compare::calculate_ast_distance;

use crate::formatting::to_tool_result;
use crate::schemas::{CompareCodeInput, CompareFilesInput, ComparisonResult};
use crate::security::read_safe_utf8_file;
use crate::server::ToposServer;

fn failed_comparison(error: String, source_valid: bool, target_valid: bool) -> ComparisonResult {
    ComparisonResult {
        raw_distance: 0.0,
        normalized_distance: 0.0,
        similarity: 0.0,
        operations: Default::default(),
        source_valid,
        target_valid,
        warnings: Vec::new(),
        error: Some(error),
    }
}

pub(crate) fn render_comparison_md(r: &ComparisonResult) -> String {
    if let Some(err) = &r.error {
        return format!("**Error:** {err}");
    }
    let mut lines = vec![
        format!("**Normalized distance:** {:.3}", r.normalized_distance),
        format!("**Similarity:** {:.3}", r.similarity),
        format!("**Raw distance:** {:.1}", r.raw_distance),
        format!(
            "**Validity:** source={}, target={}",
            r.source_valid, r.target_valid
        ),
    ];
    if !r.operations.is_empty() {
        let mut pairs: Vec<_> = r.operations.iter().collect();
        pairs.sort_by(|a, b| a.0.cmp(b.0));
        let ops = pairs
            .iter()
            .map(|(k, v)| format!("{k}={v}"))
            .collect::<Vec<_>>()
            .join(", ");
        lines.push(format!("**Operations:** {ops}"));
    }
    lines.join("\n")
}

fn compare_code_impl(params: &CompareCodeInput) -> CallToolResult {
    let src = ProgramMorphism::new(&params.source_code, &params.language);
    let tgt = ProgramMorphism::new(&params.target_code, &params.language);

    if !(src.is_valid() && tgt.is_valid()) {
        let model = failed_comparison(
            "Failed to parse one or both code snippets.".to_string(),
            src.is_valid(),
            tgt.is_valid(),
        );
        return to_tool_result(&model, render_comparison_md(&model));
    }

    let (Some(src_ast), Some(tgt_ast)) = (src.ast.as_ref(), tgt.ast.as_ref()) else {
        let model = failed_comparison(
            "Failed to parse one or both code snippets.".to_string(),
            src.is_valid(),
            tgt.is_valid(),
        );
        return to_tool_result(&model, render_comparison_md(&model));
    };

    let result = calculate_ast_distance(src_ast, tgt_ast);
    let model = ComparisonResult {
        raw_distance: result.raw_distance as f64,
        normalized_distance: result.normalized_distance,
        similarity: 1.0 - result.normalized_distance,
        operations: result
            .operations
            .into_iter()
            .map(|(k, v)| (k, v as i64))
            .collect(),
        source_valid: true,
        target_valid: true,
        warnings: Vec::new(),
        error: None,
    };
    to_tool_result(&model, render_comparison_md(&model))
}

#[tool_router(router = compare_router, vis = "pub(crate)")]
impl ToposServer {
    /// Compute the AST (tree-edit) distance between two source-code
    /// strings.
    ///
    /// Read-only and idempotent; parses both snippets in memory, never
    /// writes or scores. Use for clone detection or to measure refactor
    /// impact; the `topos_assess_*` tools already fold this in as an
    /// anti-gaming check, so call it directly only for the raw number.
    /// Returns a ComparisonResult: `normalized_distance` in [0, 1],
    /// `similarity` (= 1 - it), `raw_distance`, an `operations` edit-count
    /// map, and `source_valid`/`target_valid` (`error` set if either fails
    /// to parse).
    #[tool(
        name = "topos_compare_code",
        annotations(
            title = "Topos Structural Comparison",
            read_only_hint = true,
            destructive_hint = false,
            idempotent_hint = true,
            open_world_hint = false
        )
    )]
    pub fn topos_compare_code(
        &self,
        Parameters(params): Parameters<CompareCodeInput>,
    ) -> CallToolResult {
        compare_code_impl(&params)
    }

    /// Compute the AST (tree-edit) distance between two source files on
    /// disk.
    ///
    /// Read-only; parses both files, never writes or scores. Use for clone
    /// detection or refactor impact; use `topos_assess_*` for a quality
    /// verdict. Returns a ComparisonResult (see `topos_compare_code`).
    #[tool(
        name = "topos_compare_files",
        annotations(
            title = "Topos Structural Comparison",
            read_only_hint = true,
            destructive_hint = false,
            idempotent_hint = true,
            open_world_hint = false
        )
    )]
    pub fn topos_compare_files(
        &self,
        Parameters(params): Parameters<CompareFilesInput>,
    ) -> CallToolResult {
        let source_text = match read_safe_utf8_file(&params.source) {
            Ok(text) => text,
            Err(err) => {
                let model = failed_comparison(format!("Source file error: {err}"), false, false);
                return to_tool_result(&model, render_comparison_md(&model));
            }
        };
        let target_text = match read_safe_utf8_file(&params.target) {
            Ok(text) => text,
            Err(err) => {
                let model = failed_comparison(format!("Target file error: {err}"), true, false);
                return to_tool_result(&model, render_comparison_md(&model));
            }
        };
        compare_code_impl(&CompareCodeInput {
            source_code: source_text,
            target_code: target_text,
            language: "python".to_string(),
        })
    }
}
